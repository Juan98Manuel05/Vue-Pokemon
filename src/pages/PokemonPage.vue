<template>
    <h1 v-if="!pokemon"> Espere por favor...</h1>
    
    <div v-else-if="pokemon">
        <h1>¿ Quien es este pokémon?</h1>

        <!-- IMG -->
        <PokemonPicture 
            :pokemon-id="pokemon.id" 
            :show-pokemon="showPokemon"/>

        <!-- OPCIONES -->
        <PokemonOptions            
            :pokemons="pokemonsArr" 
            @selection-pokemon="checkAnswer"
            />

        <template v-if="showAnswer">
            <h2 class="fade-in"> {{ message }} </h2>
            <button @click="newGame">
                nuevo juego
            </button>
        </template>
    </div>

</template>

<script>

import PokemonPicture from '@/components/PokemonPicture'
import PokemonOptions from '@/components/PokemonOptions'

import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {
    components:{
        PokemonPicture, PokemonOptions
    },
    data(){
        return{
            pokemonsArr: [],
            pokemon: null,
            showPokemon: false,
            showAnswer: false,
            message: ''
        }
    },
    methods:{
        async mixPokemonArray(){
            this.pokemonsArr = await getPokemonOptions()
            
            const rndInt = Math.floor(Math.random() * 4 )

            this.pokemon = this.pokemonsArr[rndInt]
        },
        checkAnswer( SelectedId ){

            this.showPokemon = true 
            this.showAnswer = true

            if( SelectedId === this.pokemon.id){
                this.message = `Correcto, ${ this.pokemon.name }`
            }else{
                this.message = `Oops era, ${ this.pokemon.name }`
            }

        },
        async newGame(){
            this.showPokemon = false
            this.showAnswer = false
            this.pokemonsArr = []
            this.pokemon =  null
            await this.mixPokemonArray()
        }
    },
    mounted(){       
        this.mixPokemonArray()
    }
}
</script>