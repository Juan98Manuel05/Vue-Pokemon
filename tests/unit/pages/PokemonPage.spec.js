import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from '../mocks/pokemons.mock'
import PokemonPicture from '@/components/PokemonPicture'
import PokemonOptions from '@/components/PokemonOptions'



describe('PokemonPage component', () => {
    
    let wrapper 
    
    beforeEach(() => {
        wrapper = shallowMount( PokemonPage )
    })
    
    test('debe de llamar el mixedPokemonArr al montar', () => {
        const mixedPokemonArrSpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray' )
        shallowMount(PokemonPage)
        
        expect( mixedPokemonArrSpy ).toHaveBeenCalled()
    })

    test('debe hacer match con el snapshot cuando se cargan los pokemons', () => {
        
        // console.log(pokemons)
        const wrapper = shallowMount( PokemonPage, {
            data(){
                return {
                    pokemonsArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })
        expect( wrapper.html() ).toMatchSnapshot()
    })

    test('debe mostrar los componentes de pokemonPicture y pokemonOptions', () => {    
        
        const wrapper = shallowMount( PokemonPage, {
            data(){
                return {
                    pokemonsArr: pokemons,
                    pokemon: pokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        const picture = wrapper.find('pokemon-picture-stub')
        const options = wrapper.find('pokemon-options-stub')
        // PokemonPicture debe existir
        expect(picture.exists()).toBeTruthy();
        // // PokemonOptions debe existir
        expect(options.exists()).toBeTruthy();

        // PokemonPicture attributte === 1
        expect(picture.attributes().pokemonid).toBe('1')
        // PokemonOptions attribute pokemons to be True
        expect(options.attributes('pokemons')).toBeTruthy()
    })

    test('pruebas con checkaAnswer', async() => {
        const wrapper = shallowMount( PokemonPage, {
            data(){
                return {
                    pokemonsArr: pokemons,
                    pokemon: pokemons[1],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(2)
        
        expect( wrapper.find('h2').exists() ).toBeTruthy()
        expect( wrapper.vm.showPokemon ).toBeTruthy()
        expect( wrapper.find('h2').text() ).toBe(`Correcto, ${pokemons[1].name}`)

        await wrapper.vm.checkAnswer(5)
        
        expect(wrapper.vm.message).toBe(`Oops era, ${pokemons[1].name}`)


    })
})