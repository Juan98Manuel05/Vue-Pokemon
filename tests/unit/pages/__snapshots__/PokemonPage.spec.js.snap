// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component debe hacer match con el snapshot cuando se cargan los pokemons 1`] = `
<div>
  <h1>¿ Quien es este pokémon?</h1><!-- IMG -->
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub><!-- OPCIONES -->
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;
